### ISSUE TITLE
Is the API down?  Thanks!

### ISSUE THREAD

**Issue by redesigned** [MartinsOnuoha/countriesNowAPI#88]
https://countriesnow.space/api/v0.1/countries/states

![Screenshot 2022-02-17 164451](https://user-images.githubusercontent.com/351192/154595989-2e89558c-cf48-460d-95df-dcb3dc031cdd.jpg)

**Comment by MartinsOnuoha** [comment]
Hi, sorry about this, it's not down, you can check the endpoint again.

**Comment by xXAvoraXx** [comment]
same question

**Comment by thomascjenks** [comment]
The endpoint seems to be broken again.

![image](https://user-images.githubusercontent.com/100152067/154995048-c2c46836-f13c-431f-8133-02ecb1909948.png)

**Comment by MartinsOnuoha** [comment]
@thomascjenks @xXAvoraXx  can you give a bit more detail how you got the error?

**Comment by xXAvoraXx** [comment]
> @thomascjenks @xXAvoraXx can you give a bit more detail how you got the error?

problem fixed.

**Comment by TimAagaard** [comment]
@MartinsOnuoha Same error is still happening today.  Was happening intermittently the last few days as well.

https://countriesnow.space/ shows "Application Error" from above responses.

Trying to hit the endpoints results in an HTTP 503 error code

**Comment by MartinsOnuoha** [comment]
Hey guys, @TimAagaard , @xXAvoraXx @redesigned  the API itself is fine. I believe someone might be using it wrongly and perhaps making heavy requests or had an infinite loop bug that caused multiple requests for about 30 minutes. Then everything came back up. 
I'd set up another monitoring tool (might be expensive) to pick out the faulting source IP or client. and blacklist it. 


<img width="347" alt="CleanShot 2022-02-25 at 8 41 36@2x" src="https://user-images.githubusercontent.com/29035007/155675015-c7dcf7df-1fc3-4f99-b4e2-31aa3ccf80be.png">

If this makes sense, can we consider this issue resolved?

**Comment by nataliiaprecoro** [comment]
It's still down. The issue can not be resolved.

**Comment by MartinsOnuoha** [comment]
It's back up now, there's an underlying problem, and it has to do with the fact that we run on heroku's free dyno, and some folks might be misusing the service.

**Comment by nataliiaprecoro** [comment]
<img width="1680" alt="image (13)" src="https://user-images.githubusercontent.com/100600075/156021360-1785ff9a-9cf2-4c3b-8d19-5da2efcf8332.png">


It seems like it doesn't work.

**Comment by MartinsOnuoha** [comment]
Hey @nataliiaprecoro The API is up now

**Comment by TimAagaard** [comment]
@MartinsOnuoha I would recommend implementing fail2ban to monitor your log files and ban IPs if they go over a threshold...or build it into the source code.

Something like 200 requests in a minute by same IP gives a 5 minute ban or something like that.  Something that will prevent abuse and keep the server up for everyone while rate limiting abusers and being a mild inconvenience...a 5 minute ban for heavy or abusive users keeps the service up for everyone but forces heavy or abusive users out of commission for a few minutes.

I am willing to assist in setting up fail2ban or any code changes needed but would need to know how it's operating in heroku...looks like you have a master json pulled from an environment variable and resistance so...I doubt I could get this working locally without some help.

**Comment by MartinsOnuoha** [comment]
Hey Tim @TimAagaard, that sounds like a great idea. I had thought of setting up a rate limiter, however I also have to consider the most cost effective. I had a look at fail2ban but I’m not too sure if it would be right for us at this time, I figured we could use a 3rd party library like [express-rate-limit](https://www.npmjs.com/package/express-rate-limit).

Would you like to work on integrating this library and applying it to all requests?

**Comment by xXAvoraXx** [comment]
The API is great, I love it, is it possible to add symbols to currencies? For example, the symbol for the dollar is '$'.

**Comment by TimAagaard** [comment]
@MartinsOnuoha Sure thing!  I implemented express-rate-limit last night real quick for all /api routes and I'm seeing the headers in the response...so it should be working but I'll test it out tonight.  I put the rate limiter config object in /config.  You'll have to tune that to what the server can handle and if you're behind a load balancer you'll have to increment a variable by the number of proxies.

Expect a PR in a couple hours when I get off work.

**Comment by MartinsOnuoha** [comment]
> The API is great, I love it, is it possible to add symbols to currencies? For example, the symbol for the dollar is '$'.
 
 hey @xXAvoraXx I’m glad it’s been useful for you, symbols does sounds like an interesting idea, here’s what you can do,  create a new issue about this so we can all follow up on the possibilities of having this feature, and also not lose the conversation within this issue.

**Comment by MartinsOnuoha** [comment]
> @MartinsOnuoha Sure thing! I implemented express-rate-limit last night real quick for all /api routes and I'm seeing the headers in the response...so it should be working but I'll test it out tonight. I put the rate limiter config object in /config. You'll have to tune that to what the server can handle and if you're behind a load balancer you'll have to increment a variable by the number of proxies.
> 
> Expect a PR in a couple hours when I get off work.

Awesome!! 🏄🏼‍♂️, I’ll be on standby, let me know how it goes

**Comment by xXAvoraXx** [comment]
> > @MartinsOnuoha Sure thing! I implemented express-rate-limit last night real quick for all /api routes and I'm seeing the headers in the response...so it should be working but I'll test it out tonight. I put the rate limiter config object in /config. You'll have to tune that to what the server can handle and if you're behind a load balancer you'll have to increment a variable by the number of proxies.
> > Expect a PR in a couple hours when I get off work.
> 
> Awesome!! 🏄🏼‍♂️, I’ll be on standby, let me know how it goes

API down again.

**Comment by TimAagaard** [comment]
@MartinsOnuoha PR sent and pending your review.

All unit tests still passing.  Confirmed rate-limit by holding ctrl+shift+r in browser until I got the rate-limit response.

You might have to alter config/rateLimit.js::numberOfProxies if you're behind a load-balancer or 2.  Just increment that value by 1 until it stops breaking.  Can be tested further on your server by adding the route listed here: https://www.npmjs.com/package/express-rate-limit#troubleshooting-proxy-issues
